### DISCUSSION TITLE
Option to ignore optional dependencies

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/vercel#4960] [Ideas] [answered]
#### Cleaner code
`package.json` does not give much freedom to classify dependencies depending on the context. Mainly, `development` is considered as "anything that is not production". However, in practice, it mixes 2 very different kind of deps: build dependencies (webpack, babel and friends), and actual develomment tooling that the app itself do not really need (cypress, jest, storybook and family).

#### Less risk of failing build

Technically, development tooling is actually "optional" dependencies. It is not need to build the app, so in my opinion we should avoid installing those on distant server if we don't need them. For instance, most of the time, Storybook has nothing to do on production environment.

For instance, I just hit the scenario where Cypress failed to download and install on Vercel... but I don't need Cypress at all for my app to run.

On a dockerized version of my app, putting non-vital development tooling in optionalDependencies saved around 600mo (from 1.9go to 1.3go) and minutes of build time.

### Palliative

Currently, the only way I found to not install optional deps was to force installing only production deps with `NPM_ONLY_PRODUCTION` and then, during build, run `yarn install --ignore-optional && yarn run build` to get the dev deps too.

~~I am currently testing it, not sure it will even work~~ It works ok. See https://github.com/vercel/vercel/discussions/4958 for the discussion I've opened in the meantime to find such a palliative solution.

### Feature request

`NPM_IGNORE_OPTIONAL` or similar would be great to be able to tweak the `yarn install` step. Or even being able to customize it?

**Comment by styfle** [comment, marked as answer]
Hi @eric-burel 

Thanks for the suggestion!

We are considering adding a new feature to allow users to customize the Install Command, similar to the way you can customize the Build Command.

This is being tracked internally as Story 5108 and I believe it would work for your use case, thanks!

**Reply by eric-burel** [reply]
Great news thanks!

**Reply by eric-burel** [reply]
Maybe one technical question: how a custom install could affect performance? One big limitation I hit with my custom Dockerfile (not hosted on Vercel) is that I can't find a way to reliably prevent a new "yarn install" during image build. This leads to a systematic reinstall even when the package list evolves, which in turns is maybe half of our deployment time (multiple minutes). 
I hope to avoid such issues when using Vercel.

**Reply by paulogdm** [reply]
We've released the ability to customize the install command, as requested.

![DeepinScreenshot_select-area_20201112075330](https://user-images.githubusercontent.com/7690649/98934713-492e6f00-24c1-11eb-9f8b-10f021700e94.png)

**Reply by eric-burel** [reply]
@paulogdm You might want to add those options on the "Import project" view as well, which currently list only build command and output directory option.

**Reply by paulogdm** [reply]
Absolutely. We have an internal story to track this.

**Reply by eric-burel** [reply]
Awesome thanks a lot :)

**Comment by raghulkrishna** [comment]
any update on this to ignore certain dependencies ?
